mypy/checkexpr.py

```diff
@@ -2066,9 +2066,63 @@ def infer_lambda_type_using_context(self, e: LambdaExpr) -> Tuple[Optional[Calla
+        """Check arguments in a super(...) call."""
+        if ARG_STAR in e.call.arg_kinds:
+            self.chk.fail('Varargs not supported with "super"', e)
+        elif e.call.args and set(e.call.arg_kinds) != {ARG_POS}:
+            self.chk.fail('"super" only accepts positional arguments', e)
+        elif len(e.call.args) == 1:
+            self.chk.fail('"super" with a single argument not supported', e)
+        elif len(e.call.args) > 2:
+            self.chk.fail('Too many arguments for "super"', e)
+        elif self.chk.options.python_version[0] == 2 and len(e.call.args) == 0:
+            self.chk.fail('Too few arguments for "super"', e)
+        elif len(e.call.args) == 2:
+            type_obj_type = self.accept(e.call.args[0])
+            instance_type = self.accept(e.call.args[1])
+            if isinstance(type_obj_type, FunctionLike) and type_obj_type.is_type_obj():
+                type_info = type_obj_type.type_object()
+            elif isinstance(type_obj_type, TypeType):
+                item = type_obj_type.item
+                if isinstance(item, AnyType):
+                    # Could be anything.
+                    return
+                if not isinstance(item, Instance):
+                    # A complicated type object type. Assume it's fine.
+                    # TODO: Do something more clever here.
+                    return
+                type_info = item.type
+            elif isinstance(type_obj_type, AnyType):
+                return
+            else:
+                self.chk.fail('First argument for "super" must be a type object', e)
+                return
+
+            if isinstance(instance_type, (Instance, TupleType, TypeVarType)):
+                if isinstance(instance_type, TypeVarType):
+                    # Needed for generic self.
+                    instance_type = instance_type.upper_bound
+                    if not isinstance(instance_type, (Instance, TupleType)):
+                        # Too tricky, give up.
+                        self.chk.fail('Unsupported argument 2 for "super"', e)
+                        return
+                if isinstance(instance_type, TupleType):
+                    # Needed for named tuples and classes derived from Tuple[...].
+                    instance_type = instance_type.fallback
+                if type_info not in instance_type.type.mro:
+                    self.chk.fail('Argument 2 for "super" not an instance of argument 1', e)
+            elif isinstance(instance_type, TypeType) or (isinstance(instance_type, FunctionLike)
+                                                         and instance_type.is_type_obj()):
+                # TODO: Check whether this is a valid type object here.
+                pass
+            elif not isinstance(instance_type, AnyType):
+                self.chk.fail('Unsupported argument 2 for "super"', e)
+
     def analyze_super(self, e: SuperExpr, is_lvalue: bool) -> Type:
```

Comment:
I remember looking at the function ``analyze_super`` and something worried me. Namely, look at the error:
``super() requires at least one positional argument`` several lines below (can't comment there), it is misleading since it can appear on Python 3 as well but we support ``super()`` without arguments. Also the logic there seems a bit wrong.

Reply:
The message refers to the enclosing function not having any positional arguments. Updated the wording. If it's still incorrect, we can create a new issue about it, since this looks like a pretty marginal scenario. The logic is kind of convoluted though -- most of the for loop body is executed at most once, which is not immediately obvious.

Reply:
You already have a list of things still left to fix in the PR description, if you feel this deserves a separate issue, then OK, otherwise you can just combine this with first item in your list.

Reply:
It's still unclear to me what the issue is (other than the code being hard to understand). Can you create a new issue (or add a comment to an existing relevant issue) about this?